foo.js

```diff
@@ -0,0 +1,182 @@
+import Alt from './'
+import React from 'react'
+import Render, { connect } from './utils/Render'
+import axios from 'axios'
+
+const alt = new Alt()
+
+const actions = alt.generateActions('yes', 'no')
+
+const UserStore = alt.createStore(function () {
+  this.state = { users: [] }
+
+  this.bindAction(actions.yes, users => this.setState({ users }))
+//  this.bindAction(actions.yes, users => console.log('data', users))
+
+  this.exportPublicMethods({
+    getUsers: () => this.state.users,
+
+    fetchUsers: (user, repo) => {
+      return this.fetch({
+        remote() {
+          const url = `https://api.github.com/repos/${user}/${repo}/stargazers`
+          return axios({ url }).then(response => response.data)
+        },
+
+        success: actions.yes,
+        error: actions.no
+      })
+    }
+  })
+}, 'UserStore')
+
+@connect({
+  propTypes: {
+    user: React.PropTypes.string.isRequired,
+    repo: React.PropTypes.string.isRequired
+  },
+
+  listenTo() {
+    return [UserStore]
+  },
+
+  resolveAsync(props, context) {
+    return context.resolve(UserStore.fetchUsers, props.user, props.repo)
+//    return UserStore.fetchUsers(props.user, props.repo)
+  },
+
+//  reduceProps(props, context) {
+//    return {
+//      users: UserStore.getUsers()
+//    }
+//  },
+
+//  loading() {
+//    return <div>Loading...</div>
+//  },
+
+  failed(props, context) {
+    return <div>Uh oh</div>
+  }
+})
+class Stargazers extends React.Component {
+  static propTypes = {
+    users: React.PropTypes.array
+  }
+
+  render() {
+    return (
+      <section>
+        <div className="card">
+          <div className="card-content">
+            <span className="card-title deep-purple-text">Stargazers</span>
+            <p>
+              People who have starred this project
+            </p>
+          </div>
+        </div>
+
+        {this.renderUsers(this.props.users)}
+      </section>
+    )
+  }
+
+  renderUser(user) {
+    return (
+      <div key={user.id} style={{ display: 'inline-block' }}>
+        <img src={user.avatar_url} alt="" width={32} height={32} />
+        <br />
+        {user.login}
+      </div>
+    )
+  }
+
+  renderUsers(users) {
+    return (
+      <div>
+        {users.map(user => (
+          <div key={user.id}>
+            {this.renderUser(user)}
+          </div>
+        ))}
+      </div>
+    )
+  }
+}
+
+@connect({
+  propTypes: {
+    user: React.PropTypes.string.isRequired,
+    repo: React.PropTypes.string.isRequired
+  },
+
+  listenTo() {
+    return [UserStore]
+  },
+
+  resolveAsync(props, context) {
+    return context.resolve(UserStore.fetchUsers, props.user, props.repo)
```

Comment:
What is this `context.resolve()` magic? Well, if there's an in-flight request then it just uses that promise instead of firing off two requests for the same fetch.

How does it work?

Calls to fetchUsers with the same arguments will be "cached" per context instance and only for a single render lifecycle. If a component's state changes the cache is blown away so this doesn't affect client side re-renders.